File: packages/luqma_core/test/merchant_sales_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:luqma_core/luqma_core.dart';

void main() {
  final cairoNow = DateTime(2026, 9, 12, 14, 0); // Saturday 2pm

  OrderLine line(String id, String name, int qty, int price) => OrderLine(
        itemId: id,
        name: name,
        quantity: qty,
        unitPrice: price,
      );

  Order makeOrder({
    required String id,
    String merchantId = 'm1',
    int subtotal = 10000,
    int deliveryFee = 1500,
    OrderStatus status = OrderStatus.delivered,
    OrderActor? cancelledBy,
    DateTime? placedAt,
    List<OrderLine> items = const [],
  }) {
    return Order(
      id: id,
      cityId: 'edku',
      orderNumber: 1,
      customerName: 'عميل',
      customerPhone: '01000000000',
      merchantId: merchantId,
      merchantName: 'مطعم',
      zoneId: 'z1',
      type: OrderType.instant,
      items: items.isEmpty
          ? [line('i1', 'وجبة', 1, subtotal)]
          : items,
      pricing: OrderPricing(
        subtotal: subtotal,
        deliveryFee: deliveryFee,
        total: subtotal + deliveryFee,
      ),
      status: status,
      cancelledBy: cancelledBy,
      placedAt: placedAt ?? cairoNow,
    );
  }

  group('MerchantSales.fromJson', () {
    test('decodes 64-bit numbers safely as num', () {
      final json = {
        'days': 7,
        'orders': 3,
        'sales': 36000,
        'average': 12000,
        'cancelledByCustomer': 1,
        'cancelledByMerchant': 0,
        'returned': 1,
        'byDay': [
          {'day': '2026-09-06', 'orders': 2, 'sales': 30000},
          {'day': '2026-09-07', 'orders': 0, 'sales': 0},
        ],
        'topItems': [
          {'itemId': 'i1', 'name': 'سمك', 'quantity': 4},
        ],
      };

      final sales = MerchantSales.fromJson(json);

      expect(sales.days, 7);
      expect(sales.orders, 3);
      expect(sales.sales, 36000);
      expect(sales.average, 12000);
      expect(sales.cancelledByCustomer, 1);
      expect(sales.cancelledByMerchant, 0);
      expect(sales.returned, 1);
      expect(sales.unfulfilled, 2);
      expect(sales.byDay.length, 2);
      expect(sales.byDay.first.day, '2026-09-06');
      expect(sales.byDay.first.orders, 2);
      expect(sales.byDay.first.sales, 30000);
      expect(sales.topItems.length, 1);
      expect(sales.topItems.first.name, 'سمك');
      expect(sales.topItems.first.quantity, 4);
    });
  });

  group('MerchantSales.of (in-memory computation)', () {
    test('counts delivered food sales only, ignoring delivery fees', () {
      final fish = [line('i1', 'سمك', 2, 5000)];
      final rice = [line('i2', 'رز', 1, 6000)];

      final orders = [
        makeOrder(
          id: 'o1',
          subtotal: 10000,
          deliveryFee: 2000,
          placedAt: cairoNow,
          items: fish,
        ),
        makeOrder(
          id: 'o2',
          subtotal: 20000,
          deliveryFee: 1500,
          placedAt: cairoNow.subtract(const Duration(days: 1)),
          items: fish,
        ),
        makeOrder(
          id: 'o3',
          subtotal: 6000,
          deliveryFee: 1000,
          placedAt: cairoNow.subtract(const Duration(days: 2)),
          items: rice,
        ),
        // Cancelled orders
        makeOrder(
          id: 'o4',
          subtotal: 9000,
          status: OrderStatus.cancelled,
          cancelledBy: OrderActor.customer,
          placedAt: cairoNow,
        ),
        makeOrder(
          id: 'o5',
          subtotal: 9000,
          status: OrderStatus.cancelled,
          cancelledBy: OrderActor.courier,
          placedAt: cairoNow,
        ),
        // Outside the 7-day window (30 days ago)
        makeOrder(
          id: 'o6',
          subtotal: 99000,
          placedAt: cairoNow.subtract(const Duration(days: 30)),
        ),
        // Another shop
        makeOrder(
          id: 'o7',
          merchantId: 'm2',
          subtotal: 77000,
          placedAt: cairoNow,
        ),
      ];

      final sales = MerchantSales.of(
        orders,
        merchantId: 'm1',
        days: 7,
        now: () => cairoNow,
      );

      expect(sales.orders, 3);
      expect(sales.sales, 36000, reason: 'Food only (10000+20000+6000), no delivery fee');
      expect(sales.average, 12000, reason: '36000 ~/ 3');
      expect(sales.cancelledByCustomer, 1);
      expect(sales.cancelledByMerchant, 0);
      expect(sales.returned, 1, reason: 'Courier cancellation counts as returned');
      expect(sales.unfulfilled, 2);

      // byDay must cover all 7 days in order
      expect(sales.byDay.length, 7);
      expect(sales.byDay.where((d) => d.orders > 0).length, 3);
      expect(sales.byDay.where((d) => d.orders == 0).length, 4);
      expect(sales.byDay.last.day, '2026-09-12');
      expect(sales.byDay.first.day, '2026-09-06');

      // Top items
      expect(sales.topItems.length, 2);
      expect(sales.topItems[0].name, 'سمك');
      expect(sales.topItems[0].quantity, 4);
      expect(sales.topItems[1].name, 'رز');
      expect(sales.topItems[1].quantity, 1);
    });

    test('survives an all-zero week without division by zero', () {
      final sales = MerchantSales.of(
        [],
        merchantId: 'm1',
        days: 7,
        now: () => cairoNow,
      );

      expect(sales.orders, 0);
      expect(sales.sales, 0);
      expect(sales.average, 0);
      expect(sales.cancelledByCustomer, 0);
      expect(sales.cancelledByMerchant, 0);
      expect(sales.returned, 0);
      expect(sales.byDay.length, 7);
      for (final d in sales.byDay) {
        expect(d.orders, 0);
        expect(d.sales, 0);
      }
      expect(sales.topItems, isEmpty);
    });

    test('clamps days window to 1..90', () {
      final salesLow = MerchantSales.of([], merchantId: 'm1', days: 0, now: () => cairoNow);
      expect(salesLow.days, 1);
      expect(salesLow.byDay.length, 1);

      final salesHigh = MerchantSales.of([], merchantId: 'm1', days: 120, now: () => cairoNow);
      expect(salesHigh.days, 90);
      expect(salesHigh.byDay.length, 90);
    });
  });

  group('FakeMerchantSalesRepository', () {
    test('computes from seeded orders', () async {
      final orders = [
        makeOrder(id: 'o1', subtotal: 15000, placedAt: cairoNow),
      ];
      final repo = FakeMerchantSalesRepository(seed: orders, now: () => cairoNow);

      final result = await repo.getSales('m1', days: 7);
      expect(result.isOk, isTrue);
      final sales = result.valueOrThrow;
      expect(sales.orders, 1);
      expect(sales.sales, 15000);
    });

    test('returns failure when instructed', () async {
      final repo = FakeMerchantSalesRepository(
        failure: const OfflineFailure(),
        now: () => cairoNow,
      );

      final result = await repo.getSales('m1');
      expect(result is Err, isTrue);
      expect(result.failureOrNull, isA<OfflineFailure>());
    });
  });
}

File: supabase/test/local/merchant_sales.test.js
import { after, before, describe, it } from 'node:test';
import assert from 'node:assert/strict';
import { freshDatabase } from './harness.mjs';

/**
 * What a shop sold this week.
 *
 * Sales are the **food** — `pricing.subtotal`, the figure commission is charged on —
 * because when the platform delivers, the delivery fee was never the merchant's to see,
 * and a headline that includes it on some orders and not others is a number the owner
 * cannot check against their own till.
 *
 * A cancelled order is not a sale and is counted anyway, on its own, split by who
 * cancelled it: a customer changing their mind and a courier coming back from the door
 * are different problems.
 */
describe('what a shop sold', () => {
  const OWNER = '00000000-0000-0000-0000-0000000000aa';
  let db, shop, other, zone;

  const as = (uid, claims) => db.exec(`
    create or replace function auth.uid() returns uuid language sql stable
      as $fn$ select '${uid}'::uuid $fn$;
    create or replace function auth.jwt() returns jsonb language sql stable
      as $fn$ select '${JSON.stringify({ app_metadata: claims })}'::jsonb $fn$;`);

  const order = async ({ merchant, subtotal, status = 'delivered',
                         cancelledBy = null, daysAgo = 0, items = [] }) => {
    const id = (await db.query(
      `insert into orders (city_id,customer_uid,customer_name,customer_phone,
                           merchant_id,merchant_name,zone_id,type,items,pricing,status)
       values ('edku',null,'عميل','01000000000',$1,'مطعم',$2,'instant',$3::jsonb,
               jsonb_build_object('subtotal',$4::int,'total',$4::int + 1000),'placed')
       returning id`,
      [merchant, zone, JSON.stringify(items), subtotal])).rows[0].id;

    await db.query(`do $$ begin
      perform set_config('app.server_mode','on',true);
      update public.orders
         set status = '${status}',
             cancelled_by = ${cancelledBy ? `'${cancelledBy}'` : 'null'},
             placed_at = (now() at time zone 'Africa/Cairo')::date
                         - interval '${daysAgo} days' + interval '13 hours'
       where id = '${id}';
    end $$;`);
    return id;
  };

  before(async () => {
    db = await freshDatabase();
    await db.exec(`
      insert into auth.users (id) values ('${OWNER}');
      grant usage on schema auth to authenticated;
      insert into cities (id,name) values ('edku','إدكو');`);
    zone = (await db.query(
      `insert into zones (city_id,name,default_delivery_fee)
       values ('edku','الزغبي',1000) returning id`)).rows[0].id;
    const make = async (n) => (await db.query(
      `insert into merchants (city_id,type,name,zone_id,phone,status)
       values ('edku','restaurant',$1,$2,'0100','approved') returning id`,
      [n, zone])).rows[0].id;
    shop = await make('مطعم');
    other = await make('مطعم تاني');
    await db.query(
      `insert into staff (uid,scope,role,merchant_id,is_active)
       values ($1,'merchant','owner',$2,true)`, [OWNER, shop]);

    const fish = [{ itemId: 'i1', name: 'سمك', quantity: 2 }];
    const rice = [{ itemId: 'i2', name: 'رز', quantity: 1 }];

    await order({ merchant: shop, subtotal: 10000, items: fish });
    await order({ merchant: shop, subtotal: 20000, daysAgo: 1, items: fish });
    await order({ merchant: shop, subtotal: 6000, daysAgo: 2, items: rice });
    await order({ merchant: shop, subtotal: 9000, status: 'cancelled',
                  cancelledBy: 'customer' });
    await order({ merchant: shop, subtotal: 9000, status: 'cancelled',
                  cancelledBy: 'courier' });
    // Outside the window, and somebody else's.
    await order({ merchant: shop, subtotal: 99000, daysAgo: 30 });
    await order({ merchant: other, subtotal: 77000 });
  });

  after(async () => { await db?.close(); });

  const sales = async (days = 7) => {
    await as(OWNER, { role: 'owner', scope: 'merchant', merchant_id: shop });
    await db.exec('set role authenticated');
    try {
      return (await db.query(
        'select public.merchant_sales($1,$2) as s', [shop, days])).rows[0].s;
    } finally {
      await db.exec('reset role');
    }
  };

  it('counts the food it delivered, and only that', async () => {
    const s = await sales();
    assert.equal(s.orders, 3);
    assert.equal(Number(s.sales), 36000, '10000 + 20000 + 6000, no delivery fee in it');
  });

  it('never another shop’s', async () => {
    const s = await sales();
    assert.equal(Number(s.sales), 36000);
  });

  it('and nothing older than the window asked for', async () => {
    const s = await sales();
    assert.equal(s.orders, 3, 'the order thirty days back is a different month');
  });

  it('counts what was cancelled, apart, and by whom', async () => {
    const s = await sales();
    assert.equal(s.cancelledByCustomer, 1);
    assert.equal(s.returned, 1, 'a courier coming back from the door is its own problem');
    assert.equal(Number(s.sales), 36000, 'and neither is a sale');
  });

  it('averages in whole piastres', async () => {
    const s = await sales();
    assert.equal(Number(s.average), 12000);
  });

  it('gives every day in the window, including the empty ones', async () => {
    const s = await sales();
    assert.equal(s.byDay.length, 7, 'a bar chart with holes in it is a broken chart');
    assert.equal(s.byDay.filter((d) => d.orders > 0).length, 3);
  });

  it('and the dishes that sold, most first', async () => {
    const s = await sales();
    assert.equal(s.topItems[0].name, 'سمك');
    assert.equal(Number(s.topItems[0].quantity), 4, 'two orders of two');
    assert.equal(s.topItems.length, 2);
  });

  it('says nothing rather than failing for a shop with no week', async () => {
    await as(OWNER, { role: 'owner', scope: 'merchant', merchant_id: shop });
    await db.exec('set role authenticated');
    const s = (await db.query(
      'select public.merchant_sales($1,7) as s', [other])).rows[0].s;
    await db.exec('reset role');

    // `read_orders` shows this owner nothing of the other shop, so the count is zero
    // rather than an error — RLS filters, it does not refuse.
    assert.equal(s.orders, 0);
    assert.equal(Number(s.sales), 0);
    assert.deepEqual(s.topItems, []);
  });
});
